File: Iteration 1/code/CSE3063GRP12/src/features/Advisor/CourseApprovalController.java
package features.Advisor;

import core.general_providers.TerminalManager;
import core.repositories.CourseEnrollmentRepository;
import core.models.concretes.CourseEnrollment;
import core.exceptions.UnexpectedInputException;
import core.main_menu.MenuController;
import java.util.ArrayList;

public class CourseApprovalController {
    private CourseApprovalView courseApprovalView;
    private CourseEnrollmentRepository courseEnrollmentRepository;

    public CourseApprovalController(){
        courseEnrollmentRepository = new CourseEnrollmentRepository();
        courseApprovalView = new CourseApprovalView();

        handleApprovalController();
    }


    private ArrayList<CourseEnrollment> fetchPendingEnrollments() {
        return courseEnrollmentRepository.getPendingEnrollments();
    }

    private int getUserSelection() {
        System.out.printf("Type \"q\" to see the menu\nType Student Id to process course enrollment");
        ArrayList<CourseEnrollment> courseEnrollmentList = fetchPendingEnrollments();
        courseApprovalView.showPendingCourseEnrollments(courseEnrollmentList);
        String input = TerminalManager.getInstance().read();
        TerminalManager.getInstance().dispose();
        try{
            if(input.equals("q")){
                return -1;
            }
            else{
                for(int i = 0;i < courseEnrollmentList.size();i++){
                    courseEnrollmentList.get(i).getStudentId().equals(input);
                    return i;
                }
            }
            new UnexpectedInputException(input);
        }
        catch (UnexpectedInputException e){
            new UnexpectedInputException(input);
            return -2;
        }
    }

    private void navigateToApprovalCoursesSelected(CourseEnrollment courseEnrollment){}

    private void navigateToMenu(){
        return new MenuController();
    }

    private void handleApprovalController(){
        while(true) {
            int selection = getUserSelection();
            if(selection == -1){
                navigateToMenu();
                break;
            }
            else if(selection>=0){
                navigateToApprovalCoursesSelected(fetchPendingEnrollments().get(selection));
                break;
            }
        }
    }
}

File: Iteration3/code/CSE3063GRP12/src/features/advisor/course_approval/CourseApprovalController.py
from ....core.general_providers.TerminalManager import TerminalManager
from ....core.repositories.CourseEnrollmentRepository import CourseEnrollmentRepository
from ....core.exceptions.UnexpectedInputException import UnexpectedInputException
from ..approval_courses_selected.ApprovalCoursesSelectedController import ApprovalCoursesSelectedController
from CourseApprovalView import CourseApprovalView
from ...main_menu.MenuController import MenuController

class CourseApprovalController:
    def __init__(self):
        self.__courseEnrollmentRepository = CourseEnrollmentRepository()
        self.__courseApprovalView = CourseApprovalView()

        try:
            self.handleApprovalController()
        except UnexpectedInputException as e:
            self.__courseApprovalView.showErrorMessage(e)
            CourseApprovalController()

    def fetchPendingEnrollments(self):
        return self.__courseEnrollmentRepository.get_pending_enrollments()

    def navigateToApprovalCoursesSelected(self, course_enrollment):
        ApprovalCoursesSelectedController(course_enrollment)

    def navigateToMenu(self):
        MenuController()

    def getUserInput(self):
        input_value = TerminalManager.getInstance().read()
        return input_value

    def handleApprovalController(self):
        pending_enrollments = self.fetchPendingEnrollments()
        self.__courseApprovalView.showPendingCourseEnrollments(pending_enrollments)

        self.__courseApprovalView.showPromptMessage()
        selection = self.getUserInput()
        if not selection.isdigit() and selection != "q":
            raise UnexpectedInputException()
        elif selection == "q":
            self.navigateToMenu()
        else:
            index = int(selection)
            if index > len(self.__courseEnrollmentRepository.get_pending_enrollments()) or index < 0:
                raise UnexpectedInputException()
            else:
                self.navigateToApprovalCoursesSelected(self.__courseEnrollmentRepository.get_pending_enrollments()[index - 1])
